src/lib/navigation/base-navigation-mode.interface.ts

```diff
@@ -0,0 +1,160 @@
+import {EventEmitter} from '@angular/core';
+import {WizardState} from './wizard-state.model';
+import {MovingDirection} from '../util/moving-direction.enum';
+import {NavigationMode} from './navigation-mode.interface';
+
+/**
+ * Base implementation of [[NavigationMode]]
+ *
+ * @author Marc Arndt
+ */
+export abstract class BaseNavigationMode implements NavigationMode {
+  constructor(protected wizardState: WizardState) {
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public canGoToStep(destinationIndex: number): Promise<boolean> {
+    const hasStep = this.wizardState.hasStep(destinationIndex);
+
+    const movingDirection = this.wizardState.getMovingDirection(destinationIndex);
+
+    const canExitCurrentStep = (previous: boolean) => {
+      return previous && this.wizardState.currentStep.canExitStep(movingDirection);
+    };
+
+    const canEnterDestinationStep = (previous: boolean) => {
+      return previous && this.wizardState.getStepAtIndex(destinationIndex).canEnterStep(movingDirection);
+    };
+
+    const canTransitionToStep = (previous: boolean) => {
+      return previous && this.canTransitionToStep(destinationIndex);
+    };
+
+    return Promise.resolve(hasStep)
+      .then(canTransitionToStep)
+      // Apply user-defined checks at the end.  They can involve user interaction
+      // which is better to be avoided if navigation mode does not actually allow the transition
+      // (`canTransitionToStep` returns `false`).
+      .then(canExitCurrentStep)
+      .then(canEnterDestinationStep);
+  }
+
+  /**
+   * Imposes additional restrictions for `canGoToStep` in current navigation mode.
+   *
+   * The base implementation allows transition iff the given step is navigable from the navigation bar (see `isNavigable`).
+   * However, in some navigation modes `canTransitionToStep` can be more relaxed to allow navigation to certain steps
+   * by previous/next buttons, but not using the navigation bar.
+   */
+  protected canTransitionToStep(destinationIndex: number): boolean {
+    return this.isNavigable(destinationIndex);
+  }
+
+  /**
+   * @inheritDoc
+   */
+  goToStep(destinationIndex: number, preFinalize?: EventEmitter<void>, postFinalize?: EventEmitter<void>): void {
```

Comment:
`public`? I'm a fan of adding the accessibility modifiers whenever possible. This makes it much easier to understand from where a method should be accessible and from where not.

Reply:
Sorry, the project I am currently working on has had bad influence on me :) Will add `public` right now.

Reply:
Should we enable "member-access" check in tslint.json ?
https://palantir.github.io/tslint/rules/member-access/

Reply:
No problem :) Yes I think it would be a good idea to enable `member-access`. I didn't know such a property existed... If you know of other useful checks, that improve the code quality, feel free to add them too!

Reply:
Done in 54a6147